### ISSUE TITLE
Memory explosion problem

### ISSUE THREAD

**Issue by EmeryLee97** [facebookresearch/theseus#654]
## ❓ Questions and Help
Hi, I'm using Theseus to do pose graph optimization, and I I encountered two problems:
1. Every time I try to run optimization, system memory usage sharply exceeds 50 Gb, although all the variables I pass into the objective only takes a few GB of memory
2. I want to have a look at the error_metric every time I add a new edge into the pose graph, and I notice that all error_metric become 'nan' since the last error_metric exceeds 1000 
Could you please give me some hints what can cause these problems?

**Comment by EmeryLee97** [comment]
I forget to change the input type from "th.Point3" (which has no additional batch size) to "th.Variable" (which has additional batch size) in part of my code, so the batch size was hundreds, and this causes memory explosion.

**Comment by luisenp** [comment]
Ah, glad to hear that. Let me know if you have other questions, I'll try to answer more quickly next time!

**Comment by EmeryLee97** [comment]
> Ah, glad to hear that. Let me know if you have other questions, I'll try to answer more quickly next time!

Hi, thanks for your response, there's indeed a problem I'm trying to fix. I keep getting nan gradients as follows. Is there anyway to check the gradients to each intermediate variable in Theseus?

Current error metric = tensor([111.0638])
Nonlinear optimizer. Iteration: 0. Error: 111.06378936767578
Nonlinear optimizer. Iteration: 1. Error: nan
Nonlinear optimizer. Iteration: 2. Error: nan
Nonlinear optimizer. Iteration: 3. Error: nan
Nonlinear optimizer. Iteration: 4. Error: nan
Nonlinear optimizer. Iteration: 5. Error: nan
Nonlinear optimizer. Iteration: 6. Error: nan
Nonlinear optimizer. Iteration: 7. Error: nan
Nonlinear optimizer. Iteration: 8. Error: nan
Nonlinear optimizer. Iteration: 9. Error: nan
Nonlinear optimizer. Iteration: 10. Error: nan

**Comment by luisenp** [comment]
You can pass a callback as an optimizer kwarg, with keyword "end_iter_callback", which will be called at the end of each iteration [like this](https://github.com/facebookresearch/theseus/blob/main/theseus/optimizer/nonlinear/nonlinear_least_squares.py#L207). The first argument to the callback will be the optimizer itself, and then you can access the matrix of derivatives via `optimizer.linear_solver.linearization`. If you are debugging, you might consider using the dense linearization class, and then you can get the jacobians matrix with `optimizer.linear_solver.linearization.AtA`. 

The syntax will be something like (didn't test this, but should be something similar to this if it's not correct).

```python
def my_callback(optimizer, info, delta, it):
     print(optimizer.linear_solver.linearization.AtA)

     # or evaluate the current value of individual cost functions 
     with torch.no_grad():
         cost_fn = optimizer.objective.get_cost_function("my_cost_fn")
         jacobians, error = cost_fn.jacobians()

theseus_layer.forward(input_dict, optimizer_kwargs={"end_iter_callback": my_callback})
```

Hope this helps!

**Comment by EmeryLee97** [comment]
> You can pass a callback as an optimizer kwarg, with keyword "end_iter_callback", which will be called at the end of each iteration [like this](https://github.com/facebookresearch/theseus/blob/main/theseus/optimizer/nonlinear/nonlinear_least_squares.py#L207). The first argument to the callback will be the optimizer itself, and then you can access the matrix of derivatives via `optimizer.linear_solver.linearization`. If you are debugging, you might consider using the dense linearization class, and then you can get the jacobians matrix with `optimizer.linear_solver.linearization.AtA`.
> 
> The syntax will be something like (didn't test this, but should be something similar to this if it's not correct).
> 
> ```python
> def my_callback(optimizer, info, delta, it):
>      print(optimizer.linear_solver.linearization.AtA)
> 
>      # or evaluate the current value of individual cost functions 
>      with torch.no_grad():
>          cost_fn = optimizer.objective.get_cost_function("my_cost_fn")
>          jacobians, error = cost_fn.jacobians()
> 
> theseus_layer.forward(input_dict, optimizer_kwargs={"end_iter_callback": my_callback})
> ```
> 
> Hope this helps!

I print the Jacobians, and it seems when optimizing a th.SO3 variable, I can't initialize it to identity, otherwise I will get nan values when doing backpropagation (due to the .sqrt(), see link below). 
https://github.com/facebookresearch/theseus/blob/e07569138cafa2d5da3e16ff2586d2495e77817c/theseus/geometry/so3.py#L228

**Comment by luisenp** [comment]
@fantaosha Can you take a look at this? Maybe we are missing an eps here?

@EmeryLee97 Can you provide a short example that reproduces this error?

**Comment by EmeryLee97** [comment]
Sure, I firstly create a th.SE3 variable:
vertex = th.SE3(tensor=torch.tile(torch.eye(3, 4), [1, 1, 1])
and  transform its rotation part to a quaternion:
quat = vertex.rotation().to_quaternion()
then I did some quaternion multiplication and turn the result quaternion into a rotation matrix using my own function, and then did some other calculation using the rotation matrix.

**Comment by EmeryLee97** [comment]
> @fantaosha Can you take a look at this? Maybe we are missing an eps here?
> 
> @EmeryLee97 Can you provide a short example that reproduces this error?

Sure, I firstly create a th.SE3 variable:
vertex = th.SE3(tensor=torch.tile(torch.eye(3, 4), [1, 1, 1])
and transform its rotation part to a quaternion:
quat = vertex.rotation().to_quaternion()
then I did some quaternion multiplication and turn the result quaternion into a rotation matrix using my own function, and then did some other calculation using the rotation matrix.

**Comment by EmeryLee97** [comment]
> > @fantaosha Can you take a look at this? Maybe we are missing an eps here?
> > @EmeryLee97 Can you provide a short example that reproduces this error?
> 
> Sure, I firstly create a th.SE3 variable: vertex = th.SE3(tensor=torch.tile(torch.eye(3, 4), [1, 1, 1]) and transform its rotation part to a quaternion: quat = vertex.rotation().to_quaternion() then I did some quaternion multiplication and turn the result quaternion into a rotation matrix using my own function, and then did some other calculation using the rotation matrix.

This is a brief example, if you can't reproduce the error please let me know, I will give more information

**Comment by luisenp** [comment]
Hi @EmeryLee97. I believe #661 (now merged) should fix this issue. Please let me know if it doesn't.
